File: frontend/src/api/public.ts
import { request } from "@/services/request";
import { API_ENDPOINTS, compactApiParams } from "@/api/routes";

export interface PublicStats {
  in_campus_cats: number;
  neuter_rate: number;
  adopted_cats: number;
  watching_cats: number;
  total_cats: number;
}

export interface PublicSiteIntroSection {
  title: string;
  text: string;
}

export interface PublicSiteInfo {
  name: string;
  association_name: string;
  university?: string;
  slogan: string;
  intro_paragraphs: string[];
  intro_sections?: PublicSiteIntroSection[];
  highlight_note?: string;
  join_info: string;
  contact_info: string;
  feeding_tips: string[];
}

export interface PublicCatListItem {
  cat_id: string;
  name: string;
  avatar_url: string | null;
  coat_color: string;
  sex: string;
  neuter_status: string;
  status: string;
  personality_tags: string[];
  alias_summary: string | null;
}

export interface PublicCatPhoto {
  file_url: string;
  caption: string | null;
}

export interface PublicCatDetail {
  cat_id: string;
  name: string;
  aliases: string[];
  avatar_url: string | null;
  photos: PublicCatPhoto[];
  coat_color: string;
  sex: string;
  neuter_status: string;
  status: string;
  personality_tags: string[];
  story: string | null;
}

export interface PublicCatList {
  items: PublicCatListItem[];
  page: number;
  page_size: number;
  total: number;
  has_more: boolean;
}

export interface PublicCatQuery {
  keyword?: string;
  coat_color?: string;
  sex?: string;
  status?: string;
  page?: number;
  page_size?: number;
  [key: string]: unknown;
}

export type PublicPostType = "trivia" | "merch";

export interface PublicPostCard {
  post_id: string;
  post_type: PublicPostType;
  title: string;
  summary: string | null;
  cover_url: string | null;
  published_at: string | null;
}

export interface PublicPostBlock {
  block_type: string;
  text: string | null;
  image_url: string | null;
}

export interface PublicPostDetail {
  post_id: string;
  post_type: PublicPostType;
  title: string;
  summary: string | null;
  cover_url: string | null;
  published_at: string | null;
  blocks: PublicPostBlock[];
}

export interface PublicPostList {
  items: PublicPostCard[];
  page: number;
  page_size: number;
  total: number;
  has_more: boolean;
}

export interface PublicPostQuery {
  type?: PublicPostType;
  page?: number;
  page_size?: number;
  [key: string]: unknown;
}

export function getPublicStats(): Promise<PublicStats> {
  return request<PublicStats>({
    url: API_ENDPOINTS.public.stats,
    method: "GET",
  });
}

export function getPublicSiteInfo(): Promise<PublicSiteInfo> {
  return request<PublicSiteInfo>({
    url: API_ENDPOINTS.public.site,
    method: "GET",
  });
}

export function getPublicCats(query: PublicCatQuery = {}): Promise<PublicCatList> {
  return request<PublicCatList>({
    url: API_ENDPOINTS.public.cats,
    method: "GET",
    data: compactApiParams(query),
  });
}

export function getPublicCatDetail(catId: string): Promise<PublicCatDetail> {
  return request<PublicCatDetail>({
    url: API_ENDPOINTS.public.cat(catId),
    method: "GET",
  });
}

export function getPublicPosts(query: PublicPostQuery = {}): Promise<PublicPostList> {
  return request<PublicPostList>({
    url: API_ENDPOINTS.public.posts,
    method: "GET",
    data: compactApiParams(query),
  });
}

export function getPublicPostDetail(postId: string): Promise<PublicPostDetail> {
  return request<PublicPostDetail>({
    url: API_ENDPOINTS.public.post(postId),
    method: "GET",
  });
}

File: backend/app/api/v1/public.py
from fastapi import APIRouter, Query, Request

from app.core.responses import api_success
from app.modules.public import service

router = APIRouter(tags=["Public"])


@router.get("/stats", summary="Get public association stats")
def public_stats(request: Request):
    data = service.get_stats().model_dump()
    return api_success(data=data, trace_id=request.state.trace_id)


@router.get("/site", summary="Get public site/association info")
def public_site(request: Request):
    return api_success(data=service.get_site_info(), trace_id=request.state.trace_id)


@router.get("/cats", summary="Get public cat list")
def public_cats(
    request: Request,
    keyword: str | None = Query(default=None, max_length=50),
    coat_color: str | None = Query(default=None, max_length=32),
    sex: str | None = Query(default=None, max_length=16),
    status: str | None = Query(default=None, max_length=16),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=service.DEFAULT_PAGE_SIZE, ge=1, le=service.MAX_PAGE_SIZE),
):
    data = service.list_cats(
        keyword=keyword,
        coat_color=coat_color,
        sex=sex,
        status=status,
        page=page,
        page_size=page_size,
    ).model_dump()
    return api_success(data=data, trace_id=request.state.trace_id)


@router.get("/cats/{cat_id}", summary="Get public cat detail")
def public_cat_detail(request: Request, cat_id: str):
    data = service.get_cat_detail(cat_id).model_dump()
    return api_success(data=data, trace_id=request.state.trace_id)


@router.get("/posts", summary="Get public posts (trivia/merch)")
def public_posts(
    request: Request,
    type: str | None = Query(default=None, max_length=16),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=service.DEFAULT_PAGE_SIZE, ge=1, le=service.MAX_PAGE_SIZE),
):
    data = service.list_posts(post_type=type, page=page, page_size=page_size).model_dump()
    return api_success(data=data, trace_id=request.state.trace_id)


@router.get("/posts/{post_id}", summary="Get public post detail")
def public_post_detail(request: Request, post_id: str):
    data = service.get_post_detail(post_id).model_dump()
    return api_success(data=data, trace_id=request.state.trace_id)
